### ISSUE TITLE
Radeon VII memory access fault training

### ISSUE THREAD

**Issue by twuebi** [ROCm/tensorflow-upstream#534]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mint Linux, Kernel 4.15.0-52
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.13.1
- Python version: 3.65
- ROCm/MIOpen version: 2.5.27
- GPU model and memory: Radeon VII

**Describe the current behavior**

Fine-tuning [BERT](https://github.com/google-research/bert/) using [this checkpoint](https://int-deepset-models-bert.s3.eu-central-1.amazonaws.com/tensorflow/bert-base-german-cased.zip) or [that checkpoint](https://storage.googleapis.com/bert_models/2018_11_23/multi_cased_L-12_H-768_A-12.zip) on a task with a considerably large number of labels (2000) crashes with a memory access fault in 70% of the cases:

```shell
Memory access fault by GPU node-1 (Agent handle: 0x55b34b19c5b0) on address 0x7f0cc7dff000. Reason: Page not present or supervisor privilege.
Aborted (core dumped)
```
~~~shell
Memory access fault by GPU node-1 (Agent handle: 0x55fb4ce57a60) on address 0x7f6757dff000. Reason: Page not present or supervisor privilege.
Aborted (core dumped)
~~~
In contrast to #414, temperature is not the issue here, edge and memory are between 40°C and 50°C, junction around 70-80°C. 
____________________________________________________

In the other 30%, it crashes with #325:
```
InvalidArgumentError (see above for traceback): Input to reshape is a tensor with 800 values, but the requested shape has 1137887642217143831
	 [[node gradients/bert/encoder/layer_0/output/LayerNorm/moments/mean_grad/Reshape (defined at /home/tobi/bert-sticker/optimization.py:71) ]]
```

____________________________________________________

The same code runs just fine on Nvidia M60, P100 and 1070.

**Describe the expected behavior**
There should be no memory corruption issues.

**Code to reproduce the issue**

I'm looking into producing a simpler reproducer.

**Comment by twuebi** [comment]
> /cc @dagamayank for awareness. It's weird that we've also run BERT for multiple weeks and never encountered any of these issues.

@sebpuetz observes the same crashes.

**Comment by twuebi** [comment]
The time until it crashes also varies a lot. 

The last run crashed after 1 hour and 6k steps:

~~~
Memory access fault by GPU node-1 (Agent handle: 0x5583050118d0) on address 0x7f171c1ff000. Reason: Page not present or supervisor privilege.
Aborted (core dumped)
~~~

After restarting, it crashed after 30 steps:

~~~
Memory access fault by GPU node-1 (Agent handle: 0x56257cc71be0) on address 0x7fc4f29ff000. Reason: Page not present or supervisor privilege.
Aborted (core dumped)
~~~

**Comment by whchung** [comment]
+@micmelesse

@micmelesse, could you try the checkpoints?

**Comment by twuebi** [comment]
It's not connected to the checkpoints, it also happens when training from scratch.

**Comment by twuebi** [comment]
`dmesg` of two crashes:

~~~shell
[42145.304293] gmc_v9_0_process_interrupt: 11 callbacks suppressed
[42145.304298] amdgpu 0000:03:00.0: [gfxhub] no-retry page fault (src_id:0 ring:24 vmid:8 pasid:32771, for process  pid 0 thread  pid 0)
[42145.304301] amdgpu 0000:03:00.0:   in page starting at address 0x00007f171c1ff000 from 27
[42145.304303] amdgpu 0000:03:00.0: VM_L2_PROTECTION_FAULT_STATUS:0x00801030
[42145.304309] Evicting PASID 32771 queues
[42145.469445] Started evicting pasid 32771
[42145.469448] Finished evicting pasid 32771
[42252.230309] perf: interrupt took too long (4011 > 3912), lowering kernel.perf_event_max_sample_rate to 49750
[42349.843335] Signal event wasn't created because limit was reached
[42990.023970] Signal event wasn't created because limit was reached
[43168.584558] Signal event wasn't created because limit was reached
[43210.625043] amdgpu 0000:03:00.0: [gfxhub] no-retry page fault (src_id:0 ring:24 vmid:8 pasid:32771, for process  pid 0 thread  pid 0)
[43210.625046] amdgpu 0000:03:00.0:   in page starting at address 0x00007fc4f29ff000 from 27
[43210.625048] amdgpu 0000:03:00.0: VM_L2_PROTECTION_FAULT_STATUS:0x00801030
~~~

**Comment by whchung** [comment]
@micmelesse would try reproduce on Radeon VII, earlier we were running on MI parts

**Comment by micmelesse** [comment]
@twuebi Can you post the exact line that you use to do the fine tuning. Are you using the run_classifier.py, run_squad.py or your own custom code?

**Comment by twuebi** [comment]
I modified `run_classifier.py` to work for sequence labeling. The label set is ~2k labels. The dataset is not public, I am putting something together that I can share.

**Comment by twuebi** [comment]
This [archive](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/files/3358019/bert_repro.tar.gz) contains the necessary bits. (Same task, different [dataset](https://github.com/UniversalDependencies/UD_German-GSD))

___________________________________________________
I've confirmed that it also crashes on this data:

~~~shell
Memory access fault by GPU node-1 (Agent handle: 0x56249bd69990) on address 0x7f95aa9ff000. Reason: Page not present or supervisor privilege. 
Aborted (core dumped)
~~~

___________________________________________________

### Commands:

`bert_config.json` and `bert-base-german-cased-vocab.txt` are taken from [this checkpoint](https://int-deepset-models-bert.s3.eu-central-1.amazonaws.com/tensorflow/bert-base-german-cased.zip). It shouldn't matter much, I've also observed memory faults with vocab & config of the multilingual BERT.

From scratch:

~~~shell
python run_classifier.py --output_dir . --data_dir . --do_train --do_predict --do_eval --vocab_file bert-base-german-cased-vocab.txt  --bert_config_file bert_config.json --train_batch_size 48 --learning_rate 0.0002 --num_train_epochs 15 --max_seq_length 110 --do_lower_case false
~~~

Fine-tune:

~~~shell
python run_classifier.py --output_dir . --data_dir . --do_train --do_predict --do_eval --vocab_file bert-base-german-cased-vocab.txt  --bert_config_file bert_config.json --train_batch_size 48 --learning_rate 0.0002 --num_train_epochs 15 --max_seq_length 110 --do_lower_case false --init_checkpoint bert-base-german-cased 
~~~

___________________________________________________

It may be a coincidence, but I've got the impression that getting close to the memory limits leads to earlier crashes.

**Comment by micmelesse** [comment]
> edit: you may want to increase `num_train_epochs` to some larger number.

How large?

**Comment by twuebi** [comment]
50? It doesn't influence the frequency of the crashes. It just avoids the
need to restart the training if it didn't crash within num_train_epochs.

Michael Melesse <notifications@github.com> schrieb am Fr., 5. Juli 2019,
05:13:

**Comment by micmelesse** [comment]
@twuebi Ok, I will run the example for 50 epochs and see if I observe a crash. In the mean time can you use one of our standard docker images here https://hub.docker.com/r/rocm/tensorflow/ and see if you encounter the issue.

**Comment by twuebi** [comment]
> @twuebi Ok, I will run the example for 50 epochs and see if I observe a crash.

If it does not crash, over at https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/issues/325#issuecomment-493318656 @sunway513 and @whchung have a node that shows the same memory corrupting behavior as mine and @sebpuetz. It may be worth testing fine-tuning BERT there.

>  In the mean time can you use one of our standard docker images here (https://hub.docker.com/r/rocm/tensorflow) and see if you encounter the issue.

Sure, do you have a reason to believe mine or @sebpuetz setup is at fault?

**Comment by twuebi** [comment]
I tried `rocm/tensorflow:rocm2.5-tf1.13-python3`. It did not crash within the last few hours. I am, however, observing a 20% throughput difference from running it in docker to running it on host. 

@sebpuetz, on a different setup (Ubuntu + 4.18 + rocm 2.5) gets the same difference in throughput:

#### Docker:
`INFO:tensorflow:examples/sec: 62.3085`

#### Host:
`INFO:tensorflow:examples/sec: 78.5586`

The first few iterations get less throughput, these are the numbers it eventually settles on.

The difference also shows in the `AvgPwr` reading of `rocm-smi`. When running on host, `AvgPwr` spikes up to 338W. On docker, it barely touches 290W. 
______________________________

I do not think a performance reduction of 20% is acceptable, also it is unclear whether the crash does not occur because of this performance reduction or due to some software issue.

**Comment by twuebi** [comment]
@micmelesse do you have an update?

**Comment by micmelesse** [comment]
@twuebi We have observed the mem fault but it is proving hard to reproduce. We also observe the performance drop on docker. We are working on collecting HIP traces but due to the sporadic nature of the memfaults and how long it takes to complete a run this will take a while. In the meantime if you find a way to reduce the time to a memfault, we can use the traces from that to figure out a solution. Hopefully you are able to move ahead in your project using our standard docker images even though there is a performance drop. Please let us know if you observe the memfault in one of your standard docker images.

**Comment by twuebi** [comment]
@micmelesse I didn't have much time for bug chasing in the last weeks. Did you make any progress?

**Comment by micmelesse** [comment]
@twuebi No, I was unable to reproduce the memfault for a second time. You can try running the code after setting `export HIP_TRACE_API=2` to see what kernel memfaults and see if the same kernel memfaults everytime when you do it again.

**Comment by twuebi** [comment]
This has been stale for 1 1/2 months. Did you make any progress on the performance hit in the docker / the memfaults? 

Right now it seems that standard NLP-NNs like transformers are broken on AMD hardware / ROCm (cf. this issue with BERT and #325) while there is no sign of a fix or active work on the issues.

**Comment by FeepingCreature** [comment]
I seem to have ran into this with a local version of stylegan2 https://github.com/NVlabs/stylegan2 with the nvidia stuff turned off, it causes a memory access fault almost immediately unless I turn off the `tf.transpose` call in networks_stylegan2.py:110 by turning off fused_modconv.

**Comment by twuebi** [comment]
If you can, I'd suggest PyTorch, did experience much less trouble there.

**Comment by sunway513** [comment]
@twuebi , do you still see this issue using ROCm3.0 stack?

**Comment by sterlingpickens** [comment]
I am seeing something similar on my VegaFE with ROCm3.0 when I run numberfields boinc project:
https://numberfields.asu.edu/NumberFields/result.php?resultid=75544730

Opening output file ../../projects/numberfields.asu.edu_NumberFields/wu_sf3_DS-14x271_Grp632749of3932160_0_r6819622_0
Now starting the targeted Martinet search:
  Num Cvecs = 5.
    Doing Cvec 1.
Memory access fault by GPU node-1 (Agent handle: 0x2657390) on address 0x7f5c42600000. Reason: Page not present or supervisor privilege.
SIGABRT: abort called
Stack trace (9 frames):
../../projects/numberfields.asu.edu_NumberFields/GetDecics_3.05_x86_64-pc-linux-gnu__opencl_amd[0x4200b0]
/lib64/libpthread.so.0(+0x13bb0)[0x7f5d45121bb0]
/lib64/libc.so.6(gsignal+0xa3)[0x7f5d44e23013]
/lib64/libc.so.6(abort+0x114)[0x7f5d44e03548]
/usr/hsa/lib/libhsa-runtime64.so.1(+0x1a5f5)[0x7f5d44b995f5]
/usr/hsa/lib/libhsa-runtime64.so.1(+0x4f23b)[0x7f5d44bce23b]
/usr/hsa/lib/libhsa-runtime64.so.1(+0x1dd67)[0x7f5d44b9cd67]
/lib64/libpthread.so.0(+0x816c)[0x7f5d4511616c]
/lib64/libc.so.6(clone+0x3f)[0x7f5d44efc69f]

Exiting...

**Comment by sterlingpickens** [comment]
Here is another error with the same config running einstein@home gravitational wave
This wasn't present in ROCm 2.10.0. Numberfields worked on an older ROCm, but hasn't worked since 2.8.x or so.
https://einsteinathome.org/task/917690484

XLAL Error - XLALGetOpenCLProgramFromSource (/home/jenkins/workspace/workspace/EaH-GW-OpenCL-Testing/SLAVE/LIBC215/TARGET/linux-x86_64/EinsteinAtHome/source/lalsuite/lalpulsar/src/OpenCLutils.c:613): clBuildProgram failed: CL_BUILD_PROGRAM_FAILURE . Error message: 
/tmp/comgr-642450/input/CompileSource:175:42: error: passing 'REAL4 *' (aka 'float *') to parameter of type 'const REAL4 *' (aka 'const float *') changes address space of pointer
  outBSGL[j] = XLALComputeBSGL( twoF[j], twoFPerDet_local, setup );
                                         ^~~~~~~~~~~~~~~~
/tmp/comgr-642450/input/CompileSource:70:31: note: passing argument to parameter 'twoFPerDet' here
                  const REAL4 twoFPerDet[PULSAR_MAX_DETECTORS],
                              ^
/tmp/comgr-642450/input/CompileSource:194:46: error: passing 'REAL4 *' (aka 'float *') to parameter of type 'const REAL4 *' (aka 'const float *') changes address space of pointer
  outBSGLtL[j] = XLALComputeBSGLtL( twoF[j], twoFPerDet_local, maxTwoFSegPerDet_local, setup);

Oddly the einstein@home pulsar search works without issue, and amicablenumbers project works without issue.

**Comment by twuebi** [comment]
> @twuebi , do you still see this issue using ROCm3.0 stack?

Am still on 2.9, no troubles using PyTorch there, don't want to upgrade unless there's a good reason to (https://github.com/ROCmSoftwarePlatform/pytorch/issues/559). Docker is inconvenient and did slow things down (https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/issues/534#issuecomment-509302747).

**Comment by sterlingpickens** [comment]
ROCm 3.0 seems like a radical change. It deprecated some things, changed some dirs around, and is using llvm 10.0.0 which isn't even in RC1 status yet upstream. I suspect it'll take another month or two to stabilize. It seems like the compiler and/or runtime are doing something wrong.

**Comment by oleid** [comment]
Is there any news about this? I've exactly the same message on ROCm3.7 when training starts, but on  an R470 GPU.
It seems to work fine for simple architectures (if I keep the param size below 500000), but for more complicated things (something inception like) even at tiny param sizes (81,445) I get this message:

```
Memory access fault by GPU node-1 (Agent handle: 0x55d5e5566890) on address 0x9fa0c4000. Reason: Page not present or supervisor privilege.
Nearby memory map:
0x9f4e00000, 0xc00000, VRAM
0x9f5e00000, 0x400000, VRAM

PtrInfo:
        Address: 0x9f4e00000-0x9f5a00000/0x9f4e00000-0x9f5a00000
        Size: 0xc00000
        Type: 1
        Owner: 0x55d5e5566890
        CanAccess: 1
                0x55d5e5566890
        In block: 0x9f4e00000, 0xc00000
PtrInfo:
        Address: 0x9f5e00000-0x9f6200000/0x9f5e00000-0x9f6200000
        Size: 0x400000
        Type: 1
        Owner: 0x55d5e5566890
        CanAccess: 1
                0x55d5e5566890
        In block: 0x9f5e00000, 0x400000
python3: /build/hsa-rocr/src/ROCR-Runtime-rocm-3.7.0/src/core/runtime/runtime.cpp:1214: static bool rocr::core::Runtime::VMFaultHandler(hsa_signal_value_t, void*): Assertion `false && "GPU memory access fault."' failed.
```

**Comment by ppanchad-amd** [comment]
Hi @twuebi. Can you please check if you are still seeing the issue with the latest ROCm? If resolved, please close the ticket. Thanks!

**Comment by twuebi** [comment]
Hi, I no longer work with these libraries & hardware, from reading the issue, it seems that your colleagues reproduced the issue a few years back, maybe you can get in touch with them / look at what they did, a bit further up I also shared code to reproduce the issue.

**Comment by adityas-amd** [comment]
Issue is obsolete, workaround could be to use official docker image/pytorch. closing the ticket now.
